orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/providers/aws/AwsDeployStagePreProcessor.groovy

```diff
@@ -17,41 +17,118 @@
 
 package com.netflix.spinnaker.orca.clouddriver.pipeline.providers.aws
 
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.ResizeServerGroupStage
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.strategies.AbstractDeployStrategyStage
 import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.strategies.DeployStagePreProcessor
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.support.TargetServerGroupResolver
+import com.netflix.spinnaker.orca.kato.pipeline.support.ResizeStrategy
 import com.netflix.spinnaker.orca.kato.pipeline.support.StageData
 import com.netflix.spinnaker.orca.pipeline.model.Stage
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.stereotype.Component
 
+import static com.netflix.spinnaker.orca.kato.pipeline.support.ResizeStrategySupport.getSource
+
 @Component
 class AwsDeployStagePreProcessor implements DeployStagePreProcessor {
   @Autowired
   ApplySourceServerGroupCapacityStage applySourceServerGroupSnapshotStage
 
+  @Autowired
+  ResizeServerGroupStage resizeServerGroupStage
+
+  @Autowired
+  TargetServerGroupResolver targetServerGroupResolver
+
   @Override
-  List<DeployStagePreProcessor.StepDefinition> additionalSteps() {
+  List<StepDefinition> additionalSteps(Stage stage) {
+    def stageData = stage.mapTo(StageData)
+    if (stageData.strategy == "rollingredblack") {
+      // rolling red/black has no need to snapshot capacities
+      return []
+    }
+
     return [
-        new DeployStagePreProcessor.StepDefinition(
-          name: "snapshotSourceServerGroup",
-          taskClass: CaptureSourceServerGroupCapacityTask
-        )
+      new StepDefinition(
+        name: "snapshotSourceServerGroup",
+        taskClass: CaptureSourceServerGroupCapacityTask
+      )
     ]
   }
 
   @Override
-  List<DeployStagePreProcessor.StageDefinition> afterStageDefinitions() {
-    return [
-        new DeployStagePreProcessor.StageDefinition(
-          name: "restoreMinCapacityFromSnapshot",
-          stageDefinitionBuilder: applySourceServerGroupSnapshotStage,
-          context: [:]
+  List<StageDefinition> beforeStageDefinitions(Stage stage) {
+    def stageData = stage.mapTo(StageData)
+    if (shouldPinSourceServerGroup(stageData.strategy)) {
+      def resizeContext = getResizeContext(stageData)
+      return [
+        new StageDefinition(
+          name: "Pin ${resizeContext.serverGroupName}".toString(),
+          stageDefinitionBuilder: resizeServerGroupStage,
+          context: resizeContext
         )
-    ]
+      ]
+    }
+
+    return []
+  }
+
+  @Override
+  List<StageDefinition> afterStageDefinitions(Stage stage) {
+    def stageData = stage.mapTo(StageData)
+    def stageDefinitions = []
+    if (stageData.strategy != "rollingredblack") {
+      // rolling red/black has no need to apply a snapshotted capacity (on the newly created server group)
```

Comment:
this preprocessor was previously excluding rolling redblack completely ... specifically because we don't need to manipulate the capacity of the newly generated server group (the rolling red/black strategy already does it!)